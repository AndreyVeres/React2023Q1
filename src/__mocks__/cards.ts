export const cardsMock = [
  {
    artist: 'qqqqqqqqqqq',
    cmc: 5,
    colorIdentity: ['1', '2'],
    colors: ['1', '2'],
    id: '1',
    imageUrl: 'htwwwwwwwwwwwwtp',
    layout: 'eeeeeeeeeee',
    manaCost: 'rrrrrrrrrrrr',
    multiverseid: 'ttttttttttttt',
    name: 'fgffffffffff',
    number: 'vvvvvvvvvvvvv',
    originalText: 'yyyyyyyyyyyyyy',
    originalType: 'uuuuuuuuuuuuu',
    power: 'nnnnnnnnnnn',
    prinntings: ['test', 'test'],
    rarity: 'mmmmmmmmmmm',
    set: 'zzzzzzzzzzz',
    setName: 'xxxxxxxxxxx',
    subtypes: ['test', 'test'],
    text: 'aaaaaaaaaaaaaa',
    toughness: 'sssssssss',
    type: 'test',
    types: ['test', 'test'],
    variations: ['test', 'test'],
    flavor: 'pppppppppppppppppp',
  },
  {
    artist: 'test',
    cmc: 5,
    colorIdentity: ['1', '2'],
    colors: ['1', '2'],
    id: '1',
    imageUrl: 'http',
    layout: 'test',
    manaCost: 'test',
    multiverseid: 'test',
    name: 'test',
    number: 'test',
    originalText: 'test',
    originalType: 'test',
    power: 'test',
    prinntings: ['test', 'test'],
    rarity: 'test',
    set: 'test',
    setName: 'test',
    subtypes: ['test', 'test'],
    text: 'test',
    toughness: 'test',
    type: 'test',
    types: ['test', 'test'],
    variations: ['test', 'test'],
    flavor: 'test',
  },
  {
    artist: 'test',
    cmc: 5,
    colorIdentity: ['1', '2'],
    colors: ['1', '2'],
    id: '1',
    imageUrl: 'http',
    layout: 'test',
    manaCost: 'test',
    multiverseid: 'test',
    name: 'test',
    number: 'test',
    originalText: 'test',
    originalType: 'test',
    power: 'test',
    prinntings: ['test', 'test'],
    rarity: 'test',
    set: 'test',
    setName: 'test',
    subtypes: ['test', 'test'],
    text: 'test',
    toughness: 'test',
    type: 'test',
    types: ['test', 'test'],
    variations: ['test', 'test'],
    flavor: 'test',
  },
  {
    artist: 'test',
    cmc: 5,
    colorIdentity: ['1', '2'],
    colors: ['1', '2'],
    id: '1',
    imageUrl: 'http',
    layout: 'test',
    manaCost: 'test',
    multiverseid: 'test',
    name: 'test',
    number: 'test',
    originalText: 'test',
    originalType: 'test',
    power: 'test',
    prinntings: ['test', 'test'],
    rarity: 'test',
    set: 'test',
    setName: 'test',
    subtypes: ['test', 'test'],
    text: 'test',
    toughness: 'test',
    type: 'test',
    types: ['test', 'test'],
    variations: ['test', 'test'],
    flavor: 'test',
  },
  {
    artist: 'test',
    cmc: 5,
    colorIdentity: ['1', '2'],
    colors: ['1', '2'],
    id: '1',
    imageUrl: 'http',
    layout: 'test',
    manaCost: 'test',
    multiverseid: 'test',
    name: 'test',
    number: 'test',
    originalText: 'test',
    originalType: 'test',
    power: 'test',
    prinntings: ['test', 'test'],
    rarity: 'test',
    set: 'test',
    setName: 'test',
    subtypes: ['test', 'test'],
    text: 'test',
    toughness: 'test',
    type: 'test',
    types: ['test', 'test'],
    variations: ['test', 'test'],
    flavor: 'test',
  },
];
