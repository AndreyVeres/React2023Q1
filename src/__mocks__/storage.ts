interface Storage {
  [key: string]: string;
}

export class StorageMock {
  private storage: Storage;

  constructor() {
    this.storage = {};
  }
  setItem(key: string, value: string) {
    this.storage[key] = value || '';
  }
  getItem(key: string) {
    return key in this.storage ? this.storage[key] : null;
  }
  removeItem(key: string) {
    delete this.storage[key];
  }

  clear() {
    this.storage = {};
  }
}
