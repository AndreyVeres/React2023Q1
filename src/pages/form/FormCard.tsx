import React from 'react';

export interface IUserInfo {
  name: string;
  surName: string;
  dob: string;
  country: string;
  male?: string;
  file: string;
  female?: string;
}

export default function FormCard({ name, surName, dob, country, male, file, female }: IUserInfo) {
  return (
    <div>
      <h2>{`${name} ${surName}`}</h2>
      <p>{dob}</p>
      <p>{country}</p>
      <p>{male || female || ''}</p>
      <p>{file}</p>
    </div>
  );
}
