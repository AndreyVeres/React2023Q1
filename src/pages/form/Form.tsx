import React, { useState } from 'react';
import { useForm } from 'react-hook-form';
import FormCard, { IUserInfo } from './FormCard';

import './form.scss';
import { dateNotFuture } from 'utils/dateValidate';

type FormValues = {
  name: string;
  surName: string;
  dob: string;
  agree: boolean;
  file: File;
  country: string;
  gender: string;
};

export default function Form() {
  const [cards, setCards] = useState<IUserInfo[]>([]);
  const {
    register,
    handleSubmit,
    formState: { errors },
    reset,
  } = useForm<FormValues>();

  const onSubmit = (data: FormValues) => {
    const file = Object.values(data.file)[0];
    setCards((prev) => [...prev, { ...data, file: URL.createObjectURL(file) }]);

    reset();
  };

  return (
    <>
      <form className="form" onSubmit={handleSubmit(onSubmit)}>
        <div className="form__item">
          <label htmlFor="name">Name:</label>
          <input
            {...register('name', {
              required: 'required field',
              pattern: {
                value: /^[A-Z][a-z]{2,}$/,
                message: 'start with a capital letter and be at least 3 characters long',
              },
            })}
            type="text"
            id="name"
            name="name"
          />
          {errors.name && <p>{errors.name.message}</p>}
        </div>

        <div className="form__item">
          <label htmlFor="surName">SurName:</label>
          <input
            {...register('surName', {
              required: 'required field',
              pattern: {
                value: /^[A-Z][a-z]{2,}$/,
                message: 'start with a capital letter and be at least 3 characters long',
              },
            })}
            type="text"
            id="surName"
            name="surName"
          />
          {errors.surName && <p>{errors.surName.message}</p>}
        </div>

        <div className="form__item">
          <label htmlFor="country">Country:</label>
          <select {...register('country', { required: true })} id="country" name="country">
            <option value="Russia">Russia</option>
            <option value="USA">USA</option>
            <option value="UK">UK</option>
          </select>
          {errors.country ? <p>{errors.country.message}</p> : null}
        </div>

        <div className="form__item">
          <label htmlFor="dob">Date of birth:</label>
          <input
            {...register('dob', {
              required: 'required field',
              validate: { dateNotFuture },
            })}
            type="date"
            id="dob"
            name="dob"
          />
          {errors.dob && <p>{errors.dob.message}</p>}
        </div>

        <div className="form__item">
          <label htmlFor="file">Загрузите файл:</label>
          <input
            data-testid="file-input"
            {...register('file', {
              required: {
                value: true,
                message: 'file is required',
              },
            })}
            type="file"
            id="file"
            name="file"
          />
          {errors.file && <p>{errors.file.message}</p>}
        </div>

        <div className="form__item">
          <label>Выберите пол:</label>
          <div>
            <input
              {...register('gender', { required: { value: true, message: 'choose one gender' } })}
              type="radio"
              id="male"
              name="gender"
              value="male"
            />
            <label htmlFor="male">Male</label>
          </div>
          <div>
            <input
              {...register('gender', { required: { value: true, message: 'choose one gender' } })}
              type="radio"
              id="female"
              name="gender"
              value="female"
            />
            <label htmlFor="female">Female</label>
          </div>
          {errors.gender && <p>{errors.gender.message}</p>}
        </div>

        <div className="form__item">
          <label htmlFor="agree">Aggre</label>
          <input
            {...register('agree', {
              required: { value: true, message: 'you must agree to everything' },
            })}
            type="checkbox"
            id="agree"
            name="agree"
          />
          {errors.agree?.message && <p>{errors.agree.message}</p>}
        </div>

        <button data-testid="submit-button" type="submit">
          Submit
        </button>
      </form>

      <div className="cards">
        {cards.map((card, index) => (
          <FormCard {...card} key={index} />
        ))}
      </div>
    </>
  );
}
