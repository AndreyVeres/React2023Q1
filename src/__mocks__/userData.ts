import { IUserInfo } from 'pages/form/FormCard';

export const mockUserData: IUserInfo = {
  name: 'myName',
  surName: 'mySurName',
  dob: '1990-01-01',
  country: 'USA',
  female: true,
  file: 'file.png',
};
