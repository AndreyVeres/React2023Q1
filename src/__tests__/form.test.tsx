import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Form from 'pages/form/Form';
import { mockUserData } from '__mocks__/userData';
import FormCard from 'pages/form/FormCard';

describe('Form component', () => {
  it('should render all form fields correctly', () => {
    render(<Form />);
    expect(screen.getByLabelText('Name:')).toBeInTheDocument();
    expect(screen.getByLabelText('SurName:')).toBeInTheDocument();
    expect(screen.getByLabelText('Date of birth:')).toBeInTheDocument();
    expect(screen.getByLabelText('Country:')).toBeInTheDocument();
    expect(screen.getByText('Male')).toBeInTheDocument();
    expect(screen.getByText('Female')).toBeInTheDocument();
    expect(screen.getByLabelText('Загрузите файл:')).toBeInTheDocument();
    expect(screen.getByLabelText('Aggre')).toBeInTheDocument();
    expect(screen.getByRole('button', { name: 'Submit' })).toBeInTheDocument();
  });

  test('form have correcty values', () => {
    render(<Form />);
    fireEvent.change(screen.getByLabelText('Name:'), { target: { value: 'TestName' } });
    fireEvent.change(screen.getByLabelText('SurName:'), { target: { value: 'TestSurName' } });
    fireEvent.change(screen.getByLabelText('Date of birth:'), { target: { value: '1990-01-01' } });
    fireEvent.change(screen.getByLabelText('Country:'), { target: { value: 'USA' } });
    fireEvent.click(screen.getByLabelText('Male'));
    fireEvent.load(screen.getByLabelText('Загрузите файл:'), {
      target: { files: [new File([''], 'file.png', { type: 'image/png' })] },
    });
    fireEvent.click(screen.getByLabelText('Aggre'));

    expect(screen.getByLabelText('Name:')).toHaveValue('TestName');
    expect(screen.getByLabelText('SurName:')).toHaveValue('TestSurName');
    expect(screen.getByLabelText('Date of birth:')).toHaveValue('1990-01-01');
    expect(screen.getByLabelText('Country:')).toHaveValue('USA');
    expect(screen.getByLabelText('Male')).toBeChecked();
    expect(screen.getByLabelText('Female')).not.toBeChecked();
    expect(screen.getByLabelText('Загрузите файл:')).toHaveValue('');
    expect(screen.getByLabelText('Aggre')).toBeChecked();
  });

  test('render user card', () => {
    render(<FormCard {...mockUserData} />);
    expect(screen.getByTestId('name')).toHaveTextContent('myName mySurName');
    expect(screen.getByTestId('dob')).toHaveTextContent('1990-01-01');
    expect(screen.getByTestId('country')).toHaveTextContent('USA');
  });
});
