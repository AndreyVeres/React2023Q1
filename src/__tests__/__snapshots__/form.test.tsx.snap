// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders correctly 1`] = `
<div>
  <form
    class="form"
  >
    <div
      class="form__item"
    >
      <label
        for="name"
      >
        Name:
      </label>
      <input
        id="name"
        name="name"
        type="text"
      />
    </div>
    <div
      class="form__item"
    >
      <label
        for="surname"
      >
        SurName:
      </label>
      <input
        id="surname"
        name="surname"
        type="text"
      />
    </div>
    <div
      class="form__item"
    >
      <label
        for="dob"
      >
        Date of birth:
      </label>
      <input
        id="dob"
        name="dob"
        type="date"
      />
    </div>
    <div
      class="form__item"
    >
      <label
        for="country"
      >
        Country:
      </label>
      <select
        id="country"
        name="country"
      >
        <option
          value="Russia"
        >
          Russia
        </option>
        <option
          value="USA"
        >
          USA
        </option>
        <option
          value="UK"
        >
          UK
        </option>
      </select>
    </div>
    <div
      class="form__item"
    >
      <label>
        Выберите пол:
      </label>
      <div>
        <input
          id="male"
          name="gender"
          type="radio"
          value="male"
        />
        <label
          for="male"
        >
          Male
        </label>
      </div>
      <div>
        <input
          id="female"
          name="gender"
          type="radio"
          value="female"
        />
        <label
          for="female"
        >
          Female
        </label>
      </div>
    </div>
    <div
      class="form__item"
    >
      <label
        for="file"
      >
        Загрузите файл:
      </label>
      <input
        id="file"
        name="file"
        type="file"
      />
    </div>
    <div
      class="form__item"
    >
      <label
        for="agree"
      >
        Aggre
      </label>
      <input
        id="agree"
        name="agree"
        type="checkbox"
      />
    </div>
    <button
      data-testid="submit-button"
      type="submit"
    >
      Submit
    </button>
  </form>
  <div
    class="cards"
  />
</div>
`;
